runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/GrpcWindmillServerTest.java

```diff
@@ -695,8 +705,24 @@ public void onCompleted() {
     }
     stream.flush();
     stream.close();
+    isClientClosed.set(true);
 
-    Thread.sleep(100);
+    long deadline = System.currentTimeMillis() + 60_000; // 1 min
+    while (true) {
+      Thread.sleep(100);
+      int tmpErrorsAfterClose = errorsAfterClose.get();
+      int tmpErrorsBeforeClose = errorsBeforeClose.get();
+      // wait for at least 2 errors before and after
+      if (tmpErrorsAfterClose > 1 && tmpErrorsBeforeClose > 1) {
+        break;
+      }
+      if (System.currentTimeMillis() > deadline) {
+        fail(
```

Comment:
Sorry this is a bit of nit, but this check is actually verifying the test code itself, right? That is, if we get here there's likely a bug in the test code, not in the implementation? If so, it would be nice to clarify this in a comment in case this ever starts flaking.